@xen-orchestra/fs/src/utils.js

```diff
@@ -0,0 +1,45 @@
+import { createLogger } from '@xen-orchestra/log'
+import { TimeoutError } from 'promise-toolbox'
+const { info, warn } = createLogger('xo:fs:abstract')
+
+export function withTimeout(fn, timeout, { onSuccess, onTimeout, onSuccessAfterTimeout, onFailureAfterTimeout } = {}) {
+  if (typeof fn.then !== 'undefined') {
+    throw new Error('Function needs to be asynchronous.')
+  }
+
+  return function (...args) {
+    let timeoutHandle
+    let didTimeout = false
+    const timeoutError = new TimeoutError('Async call timeout limit reached')
+
+    return new Promise((resolve, reject) => {
+      timeoutHandle = setTimeout(() => {
+        didTimeout = true
+        onTimeout?.()
+        if (onSuccessAfterTimeout === undefined && onFailureAfterTimeout === undefined) {
+          reject(timeoutError)
+        }
+      }, timeout)
+      fn.apply(this, args)
+        .then(result => {
```

Comment:
```suggestion
      const promise  = fn.apply(this, args)
       if (typeof promise.then !== 'undefined') {
           throw new Error('Function needs to be asynchronous.')
       }
       promise.then(result => {
```

Reply:
on a side note, promise.then can take 2 arguments 
then(onSuccess, onFailure)
it should allow you to limit repetition in the code